### ISSUE TITLE
Loading map by sectors

### ISSUE THREAD

**Issue by securmk** [otland/forgottenserver#2174]
Loading the entire map consumes so much memory and can't refresh like Cip does (on live). Wouldn't it be better to divide the map and load it by parts like most games do? It would be easy to edit and reload on live. (Something like "/reload map ThaisCity.sector" where "ThaisCity.sector" is the file that contains the city of Thais). It could also be done Cip-like with 31x31 sectors and refreshing/cleaning them on live. On live map editing would be awesome and not loading a single HUGE map like global (or any custom) would help memory consumption.

**Comment by ranisalt** [comment]
[PyOT's sectored map](https://bitbucket.org/vapus/pyot/src/f626fab1b862b6d0ddee70e8138cca3debcb4440/game/map.py?at=default&fileviewer=file-view-default) can be a good resource for anyone trying to implement that.

It would certainly require a new map format, reusing OTBM with smaller sectors and taring the result is probably fine.

**Comment by securmk** [comment]
@ranisalt I can write the code, but I don't know how would we split a map into 31x31 sectors (RL map is is almost 35k x 35k)

**Comment by ranisalt** [comment]
I think we should store the sector dimensions and XY position at the start of the file. You can use the `otbxml2sec` from PyOT too to see how it works.

**Comment by djarek** [comment]
I think you would gain a lot more if we modified both the map file format and internal map data structures in such a way that they entire map can be mmaped into memory. Loading (and unloading when needed due to memory management pressure) would be offloaded to the kernel page management system and CPU's MMU.

**Comment by securmk** [comment]
Some months ago a friend of mine was working on a way to load house items from a simple file he called "house.db" with an structure like this:
`{201} = {32412, 31661, 7} -> {2160, 100}`
Where "201" was the house id, the {32412, 31661, 7} was the tile position where the item was placed and the {2160, 100} was the item id and count. Does this has any advantage over the traditional "tile_store"? I want to change the map structure anyway, in my opinion OTBM is deprecated and we should drop it.

**Comment by djarek** [comment]
@securmk 
By storing persistent items (a.k.a. ones that don't disappear after a restart) in a file you're essentially implementing a database. Most likely a bad one, because it most likely does not implement any of the [ACID](https://en.wikipedia.org/wiki/ACID) principles. Creating a database engine is a serious enterprise.

**Comment by securmk** [comment]
Hmm, but wouldn't it be faster than reading from the database?

@ranisalt I tried converting items.otb to json so I could convert the map but it shows me this 
`Traceback (most recent call last):
  File "/Users/MK/sec/otbxml2json.py", line 283, in <module>
    if items[id].type and items[id].type > 2: # I don't think we care for type 0 or type 1 or type 2 (aga containe$
KeyError: 30001`

I tried removing that line but then it won't launch. Any idea?

**Comment by djarek** [comment]
@securmk 
Yes it would be faster if done right. But I don't think it would be worth the effort considering the risks. What I proposed was mmaping the static map data, not the dynamic and persistent stuff and this would still be a monstrous task.

**Comment by ranisalt** [comment]
It's hardly ever a good idea to implement something if there's already a rock solid implementation available, e.g. databases and password hashes.

From what I know, the OTBM format is essentially linked, so it's not possible to read arbitrary tiles without reading everything that comes before. It would probably be better if the structure was separated between tiles and items, and every tile has a pointer to it's item list (and size?) so you could read arbitrary tiles instead of the entire map.
